packages/graphql/src/schema/validation/custom-rules/valid-types/valid-field-types.ts

```diff
@@ -17,15 +17,46 @@
  * limitations under the License.
  */
 
-import type { ASTVisitor, DirectiveNode } from "graphql";
+import type { ASTVisitor, DirectiveNode, FieldDefinitionNode } from "graphql";
+import { Kind } from "graphql";
 import type { SDLValidationContext } from "graphql/validation/ValidationContext";
+import { SCALAR_TYPES } from "../../../../constants";
 import { verifyId } from "../directives/id";
-import { verifyRelationshipFieldType } from "../directives/relationship";
 import { verifyTimestamp } from "../directives/timestamp";
 import { verifyUnique } from "../directives/unique";
 import type { ValidationFunction } from "../utils/document-validation-error";
-import { createGraphQLError, assertValid } from "../utils/document-validation-error";
+import { DocumentValidationError, createGraphQLError, assertValid } from "../utils/document-validation-error";
+import type { ObjectOrInterfaceWithExtensions } from "../utils/path-parser";
 import { getPathToNode } from "../utils/path-parser";
+import { getInnerTypeName } from "../utils/utils";
+
+function verifyRelationshipFieldType({
+    traversedDef,
+}: {
+    traversedDef: ObjectOrInterfaceWithExtensions | FieldDefinitionNode;
+}) {
+    if (traversedDef.kind !== Kind.FIELD_DEFINITION) {
+        // delegate
+        return;
+    }
+    const fieldTypeName = getInnerTypeName(traversedDef.type);
+    if (SCALAR_TYPES.includes(fieldTypeName)) {
+        const scalarTypesNotPermittedErrorMessage = `Invalid field type: Scalar types cannot be relationship targets. Please use an Object type instead.`;
+        throw new DocumentValidationError(scalarTypesNotPermittedErrorMessage, []);
+    }
+    const listTypesErrorMessage = `Invalid field type: List type relationship fields must be non-nullable and have non-nullable entries, please change type to [${getInnerTypeName(
+        traversedDef.type
+    )}!]!`;
+    if (traversedDef.type.kind === Kind.NON_NULL_TYPE) {
+        if (traversedDef.type.type.kind === Kind.LIST_TYPE) {
+            if (traversedDef.type.type.type.kind !== Kind.NON_NULL_TYPE) {
+                throw new DocumentValidationError(listTypesErrorMessage, []);
+            }
+        }
```

Comment:
What is this validating?

I see it may be hard to avoid having triple nested ifs, but maybe moving to a function may add a bit more context for the reader to see what are we validating (something like `validateListIsNonNull`?)